alica_engine/include/engine/scheduler/Job.h

```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include "engine/AlicaClock.h"
+
+#include <iostream>
+#include <chrono>
+#include <ctime>
+#include <vector>
+#include <memory>
+#include <functional>
+#include <atomic>
+
+namespace alica {
+namespace scheduler
+{
+struct Job
+{
+    int id;
+    std::function<void()> cb;
+    AlicaTime scheduledTime;
+    bool cancelled;
+    bool isRepeated;
+    bool inExecution;
+    AlicaTime repeatInterval;
+    std::vector<std::weak_ptr<Job>> prerequisites;
+
+    Job(std::function<void()> cb, std::vector<std::weak_ptr<Job>> prerequisites)
+        : cb(cb)
+        , prerequisites(prerequisites)
+        , cancelled(false)
+        , isRepeated(false)
+        , inExecution(false)
+    {
+        static std::atomic<int> sId{0};
+        id = ++sId;
+    }
+
+    bool isPrerequisite(const Job& other) const
+    {
+        for (std::weak_ptr<Job> jobWeakPtr : prerequisites) {
+            std::shared_ptr<Job> job = jobWeakPtr.lock();
+            if (job && *(job) == other) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    bool operator==(const Job& r) const
+    {
+        return id == r.id;
+    }
+
+    bool operator<(const Job& other) const
+    {
+        return isPrerequisite(other)
```

Comment:
Don't need this. since it doesn't guarantee anyway that first thing in the queue is available for execution. if need be you can just order them by scheduled time. .